lib/engine/ownable.rb

```diff
@@ -10,8 +10,19 @@ def owned_by?(entity)
       owner == entity || owner&.owner == entity || owner == entity&.owner
     end
 
+    # avoid infinite recursion for 1841
     def player
-      owner&.player? ? owner : owner&.player
+      owner_chain = [owner]
+      test_owner = owner
+      until test_owner&.player?
+        return nil unless test_owner&.owner
```

Comment:
so you're going until the current owner is a player, but are you expecting current owner to be nil? why is that?

Reply:
OK, so this is the original code for this method:
`owner&.player? ? owner : owner&.player`

**The purpose of the method is to return the player that ultimately controls (owns) the object (self) or nil if the ultimate owner is not a player.**

The way it works is if the owner is a player, then then it returns that player, if not, it recursively calls the method on the owner. For players, this is redefined to return self. For non-player ultimate owners (e.g. the bank or the share pool), this method is redefined to return nil.

I had to modify this for 1841, because 1841 1) allows corporations to own corporations and 2) "circular" ownership is legally allowed to occur (usually only in EMR scenarios). Circular ownership, (i.e. a cycle) results in infinite recursion with the original code. So I had to modify this.

Basically, I converted a recursive method into a iterative one so I could detect a cycle. If a cycle is detected I return nil. I return the owner if the owner is player. I return nil if not owned. I originally thought about coding this to be iterative for corporations and recursive otherwise, but I thought that was clunky.

Reply:
I could change L18 to be "return test_owner.player unless test_owner&.owner" so it becomes recursive if not owned.

Reply:
I've revised this. See latest commit.